File: calculator.c
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef enum { None, Op, Num, ParOpen, ParClose } TokenType;

typedef struct {
  TokenType type;
  union {
    char op;
    float num;
  };
} Token;

float expr();

char input[100];
Token tokens[100];
int tokens_len = 0;
int tok_i = 0;

void tokenize() {
  memset(tokens, 0, sizeof(tokens));
  tokens_len = 0;
  tok_i = 0;
  int n;
  for (char *p = input; *p; p += n) {
    if (isspace(*p)) {
      n = 1;
      continue;
    }

    if (isdigit(*p)) {
      tokens[tokens_len++] = (Token){.type = Num, .num = strtof(p, &p)};
      n = 0;
    } else {
      TokenType type = (*p == '(') ? ParOpen : (*p == ')') ? ParClose : Op;
      tokens[tokens_len++] = (Token){type, .op = *p};
      n = 1;
    }
  }
}

Token get_curr_tok() { return tokens[tok_i]; }

Token advance() { return tokens[tok_i++]; }

float factor() {
  Token tok = get_curr_tok();
  if (tok.type == Num)
    return advance().num;

  if (tok.type == ParOpen) {
    advance();
    float left = expr();
    return advance(), left;
  }

  exit(1);
}

float term() {
  float left = factor();
  Token tok;

  while (tok = get_curr_tok(), tok.type == Op && strchr("*/", tok.op)) {
    advance();
    float right = term();
    left = tok.op == '*' ? left * right : left / right;
  }

  return left;
}

float expr() {
  float left = term();
  Token tok;

  while (tok = get_curr_tok(), tok.type == Op && strchr("+-", tok.op)) {
    advance();
    float right = term();
    left = tok.op == '+' ? left + right : left - right;
  }

  return left;
}

void print_token(Token t) {
  if (t.type == Op) {
    printf("Op %c", t.op);
  } else if (t.type == Num) {
    printf("Num %f", t.num);
  } else {
    printf("%s", t.type == ParOpen ? "(" : ")");
  }
}

void print_tokens() {
  int i = 0;
  printf("\nTokens:[");
  while (i < tokens_len) {
    printf("\n");
    print_token(tokens[i]);
    printf(",");
    i++;
  }
  printf("\n]");
}

int main() {

  while (1) {
    printf("\ncalc> ");
    fgets(input, sizeof(input), stdin);
    tokenize();
    float res = expr();
    printf("\nresult is: %f", res);
  }

  return 0;
}

File: calculator.py
from typing import Union

INTEGER, PLUS, MINUS, MUL, DIV, LPAREN, RPAREN, EOF = (
    "INTEGER",
    "PLUS",
    "MINUS",
    "MUL",
    "DIV",
    "(",
    ")",
    "EOF",
)


class Token(object):
    def __init__(self, type, value):
        # type: (str, Union[float, str, None]) -> None
        self.type = type
        self.value = value

    def __str__(self):
        return "Token({type}, {value})".format(type=self.type, value=repr(self.value))

    def __repr__(self):
        return self.__str__()


class AST(object):
    pass


class Num(AST):
    def __init__(self, token):
        # type: (Token) -> None
        self.token = token
        self.value = token.value


class UnaryOp(AST):
    def __init__(self, op, expr):
        self.token = self.op = op
        self.expr = expr


class BinOp(AST):
    def __init__(self, left, op, right):
        # type: (Union[UnaryOp, BinOp, Num], Token, Union[UnaryOp, BinOp, Num]) -> None
        self.left = left
        self.token = self.op = op
        self.right = right


class Lexer(object):
    def __init__(self, text):
        # type: (str) -> None
        self.text = text
        self.pos = 0
        self.current_char = self.text[self.pos]

    def error(self):
        raise Exception("Invalid character")

    def advance(self):
        self.pos += 1
        if self.pos > len(self.text) - 1:
            self.current_char = None
        else:
            self.current_char = self.text[self.pos]

    def skip_whitespace(self):
        while self.current_char is not None and self.current_char.isspace():
            self.advance()

    def integer(self):
        result = ""
        while self.current_char is not None and self.current_char.isdigit():
            result += self.current_char
            self.advance()
        return int(result)

    def get_next_token(self):
        while self.current_char is not None:
            if self.current_char.isspace():
                self.skip_whitespace()
                continue

            if self.current_char.isdigit():
                return Token(INTEGER, self.integer())

            if self.current_char == "(":
                self.advance()
                return Token(LPAREN, "(")

            if self.current_char == ")":
                self.advance()
                return Token(RPAREN, ")")

            if self.current_char == "*":
                self.advance()
                return Token(MUL, "*")

            if self.current_char == "/":
                self.advance()
                return Token(DIV, "/")

            if self.current_char == "+":
                self.advance()
                return Token(PLUS, "+")

            if self.current_char == "-":
                self.advance()
                return Token(MINUS, "-")

            self.error()

        return Token(EOF, None)


class Parser(object):
    def __init__(self, lexer):
        # type: (Lexer) -> None
        self.lexer = lexer
        self.current_token = self.lexer.get_next_token()

    def error(self):
        raise Exception("Invalid syntax")

    def eat(self, token_type):
        # type: (str) -> None
        if self.current_token.type == token_type:
            self.current_token = self.lexer.get_next_token()
        else:
            self.error()

    def factor(self):
        token = self.current_token

        if token.type in (PLUS, MINUS):
            self.eat(token.type)
            node = UnaryOp(token, self.factor())
            return node
        elif token.type == LPAREN:
            self.eat(LPAREN)
            result = self.expr()
            self.eat(RPAREN)
            return result

        self.eat(INTEGER)
        return Num(token)

    def term(self):
        # type: () -> Union[UnaryOp, BinOp, Num]
        left = self.factor()

        while self.current_token.type in (MUL, DIV):
            tok = self.current_token
            self.eat(self.current_token.type)
            left = BinOp(left=left, op=tok, right=self.factor())

        return left

    def expr(self):
        # type: () -> Union[UnaryOp, BinOp, Num]
        left = self.term()

        while self.current_token.type in (PLUS, MINUS):
            tok = self.current_token
            self.eat(self.current_token.type)
            left = BinOp(left=left, op=tok, right=self.term())

        return left

    def parse(self):
        return self.expr()


class NodeVisitor(object):
    def visit(self, node):
        # type: (Union[UnaryOp, BinOp, Num]) -> Union[float, None]
        method_name = "visit_" + type(node).__name__
        visitor = getattr(self, method_name, self.generic_visit)
        return visitor(node)

    def generic_visit(self, node):
        # type: (Union[UnaryOp, BinOp, Num]) -> None
        raise Exception("No visit_{} method".format(type(node).__name__))


class Interpreter(NodeVisitor):
    def __init__(self, parser):
        # type: (Parser) -> None
        self.parser = parser

    def visit_BinOp(self, node):
        # type: (BinOp) -> Union[float, None]
        ops = {
            PLUS: lambda x, y: x + y,
            MINUS: lambda x, y: x - y,
            MUL: lambda x, y: x * y,
            DIV: lambda x, y: x / y,
        }
        left = self.visit(node.left)
        right = self.visit(node.right)
        if not left or not right:
            raise Exception("BinOp: Invalid value")

        return ops[node.op.type](left, right)

    def visit_UnaryOp(self, node):
        res = self.visit(node.expr)
        if not res:
            raise Exception("UnaryOp: Invalid value")

        return +res if node.op.type == PLUS else -res

    def visit_Num(self, node):
        # type: (Num) -> Union[float,None]
        return float(node.value) if node.value else None

    def interpret(self):
        tree = self.parser.parse()
        return self.visit(tree)


def main():
    while True:
        try:
            text = raw_input("calc> ")
        except EOFError:
            break
        if not text:
            continue
        lexer = Lexer(text)
        parser = Parser(lexer)
        interpreter = Interpreter(parser)
        result = interpreter.interpret()
        print(result)


if __name__ == "__main__":
    main()
